Report `Loss_Of_Integer_Precision` when an integer is not exactly representable as a float during conversion

radeusgd opened the pull request:
### Pull Request Description

Closes #7353 

I introduce a new type `WithAggregatedProblems`, because `WithProblems` was too simple - it only allowed to hold a `List<Problem>` but `AggregatedProblems` is more than that. Ideally we shouldn't multiply entities like this too much. We should probably unify all to use `WithAggregatedProblems` - but after starting this, I realised it will likely just take too much effort to do for this little PR. So instead, I created a follow-up task for this: #7514

### Important Notes

<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

radeusgd committed:
add error

radeusgd committed:
tests for lossy conversions

radeusgd committed:
fix

radeusgd committed:
add CSV test

radeusgd committed:
fix

radeusgd committed:
check conversion safety and store any encountered problems (no propagation yet)

radeusgd committed:
refactor class out, add conversion in Enso side

radeusgd committed:
checkpoint

radeusgd committed:
Temporarily Revert changes starting WithProblems refactor

This reverts commit 1905cc13277f836f1162ec62e3e9db04341e68aa.

radeusgd committed:
fix

radeusgd committed:
for now reverting tests for column names

radeusgd committed:
propagating warnings in cast

radeusgd committed:
propagating warnings in iif

radeusgd committed:
propagating warnings in Column.from_vector and friends

radeusgd committed:
remove unused leading-zeros logic

radeusgd committed:
propagating warnings in DelimitedReader

radeusgd commented:
I've been switching over some instances of `WithProblems` to `WithAggregatedProblems`. I think overall we want to unify all of them, but that is out of scope of this PR - to be done as #7514

radeusgd committed:
javafmt

radeusgd commented on std-bits/table/src/main/java/org/enso/table/data/column/builder/Builder.java:
```diff
@@ -96,4 +100,9 @@ public static Builder getForType(StorageType type, int size) {
 
   /** @return the current storage type of this builder */
   public abstract StorageType getType();
+
+  /** @return any problems that occurred when building the Storage. */
+  public AggregatedProblems getProblems() {
+    return AggregatedProblems.of();
+  }
```
I don't think this is a great pattern, in fact I think it is bad.

The proper solution would be to modify `seal()` to return a `WithAggregatedProblems<Storage<?>>` - that would ensure that the problems are handled throughout the codebase.

But we use `seal()` in really many places and I realised ensuring all of this is nicely handled will just make this PR far too large - thus I decided this will be better done as a separate task - #7514

GregoryTravis replied:
In general I think there's too much need to handle problem propagation explicitly in the Enso code, but I don't know enough about the big picture to suggest a solution yet.

radeusgd replied:
> In general I think there's too much need to handle problem propagation explicitly in the Enso code, but I don't know enough about the big picture to suggest a solution yet.

I'm not sure if I understand what you mean here? If anything, how error propagation in Enso code works is completely separate of how we do it in Java helpers.

I think the propagation on the Enso side works really well, although indeed there may be some places where we could want to improve them.

radeusgd commented on std-bits/table/src/main/java/org/enso/table/data/column/builder/DoubleBuilder.java:
```diff
@@ -49,13 +51,47 @@ public StorageType getType() {
+   *
+   * <p>The original LongBuilder becomes invalidated after this operation and should no longer be
+   * used.
+   */
+  static DoubleBuilder retypeFromLongBuilder(LongBuilder longBuilder) {
+    int currentSize = longBuilder.currentSize;
+    DoubleBuilder newBuilder = new DoubleBuilder(longBuilder.isMissing, longBuilder.data, currentSize);
+
+    // Invalidate the old builder.
+    longBuilder.data = null;
+    longBuilder.isMissing = null;
+    longBuilder.currentSize = -1;
+
+    // Translate the data in-place to avoid unnecessary allocations.
+    for (int i = 0; i < currentSize; i++) {
+      if (!newBuilder.isMissing.get(i)) {
+        long currentIntegerValue = newBuilder.data[i];
+        double convertedFloatValue = newBuilder.convertIntegerToDouble(currentIntegerValue);
+        newBuilder.data[i] = Double.doubleToRawLongBits(convertedFloatValue);
+      }
+    }
+
+    return newBuilder;
+  }
+
   @Override
   public void appendNoGrow(Object o) {
     if (o == null) {
       isMissing.set(currentSize++);
-    } else {
+    } else if (NumericConverter.isDecimalLike(o)){
       double value = NumericConverter.coerceToDouble(o);
       data[currentSize++] = Double.doubleToRawLongBits(value);
+    } else if (NumericConverter.isCoercibleToLong(o)) {
+      long value = NumericConverter.coerceToLong(o);
+      double converted = convertIntegerToDouble(value);
+      data[currentSize++] = Double.doubleToRawLongBits(converted);
+    } else {
+      throw new IllegalStateException("Unexpected value type when appending to a DoubleBuilder: " + o.getClass().getCanonicalName() + "." +
+          " This is a bug in the Table library.");
```
We check decimal and integer cases separately, to avoid implicit rounding-loss-of-precision and handle the loss of precision explicitly with `convertIntegerToDouble`.

radeusgd committed:
doc

radeusgd commented on std-bits/table/src/main/java/org/enso/table/data/column/builder/ObjectBuilder.java:
```diff
@@ -127,4 +128,15 @@ private void grow() {
+
+  private AggregatedProblems preExistingProblems = AggregatedProblems.of();
+
+  public void setPreExistingProblems(AggregatedProblems preExistingProblems) {
+    this.preExistingProblems = preExistingProblems;
+  }
```
This allows an ObjectBuilder after retyping to inherit any problems of the earlier builder.

radeusgd commented on std-bits/table/src/main/java/org/enso/table/data/column/storage/BoolStorage.java:
```diff
@@ -201,7 +203,7 @@ public boolean isNegated() {
   }
 
-  public Storage<?> iif(Value when_true, Value when_false, StorageType resultStorageType) {
+  public WithAggregatedProblems<Storage<?>> iif(Value when_true, Value when_false, StorageType resultStorageType) {
```
This is the pattern that should ideally be used in most places (IMO).

GregoryTravis commented on distribution/lib/Standard/Table/0.0.0-dev/src/Errors.enso:
```diff
@@ -586,6 +586,25 @@ type Conversion_Failure
+       Indicates that an automatic conversion of an integer column to a decimal
+       column is losing precision because some of the large integers cannot be
+       exactly represented by the `double` type.
+    Warning (affected_rows_count : Integer) (example_value : Integer) (example_value_converted : Decimal)
```
Could this have a row number as well?

radeusgd replied:
Hm, I would have to modify the builder a bit to be able to know the row count. Currently there is not enough context.

But do you think it's useful? Tbh I'm not sure if the `row_number` that we have in a few of these errors is useful at all - hence I didn't think of adding it.

IMO the `example_value` serves this purpose enough - the user can see one of the values that are concerned, if the row number is needed they can find it by value. Row count is useful for sure to know if this affects like one row or like 50% of the rows.

I propose to see in practice - if we start encountering these and similar warnings in practice, let's see what we do with them in practical scenarios (e.g. bookclubs, other projects) and if such row number is helpful. My current gut feeling is that example values are more useful, because they illustrate e.g. that the number in question is very large. Or e.g. for date parsing failures, ideally our error should show an example value that failed to parse - this would allow the user to compare what it is with the format they specified and see what needs to be amended in the format to make it work. Row number does not help much here. (Ofc. it is more general as from row numbers we can figure out also the values - but that requires additional actions and IMO its better to immediately show examples).

radeusgd replied:
Actually, IMO what would be more useful here is `column_name` to know with which column this warning is associated, in a multi-column table.

However, it was not trivial how to get this information from the context so I abandoned this. I'm thinking it may be worth adding it as part of the next ticket #7514. But let me know if you think that I should add it now. I imagine it should not be very problematic to do so.

GregoryTravis replied:
Agreed; the value itself is enough to find it in the data; the row number would be nice but it's not worth a big effort.

GregoryTravis commented on std-bits/table/src/main/java/org/enso/table/data/column/builder/LossOfIntegerPrecision.java:
```diff
@@ -0,0 +1,32 @@
+    return affectedRows;
+  }
+
+  void increaseAffectedRows() {
```
```suggestion
  void incrementAffectedRows() {
```

radeusgd committed:
CR

radeusgd committed:
fix a typo

radeusgd committed:
fix problem handling issue

radeusgd committed:
update a test:
we can either keep the precision of large integers and warn about unexpected rounding OR keep the +- sign of 0 (only possible in floats); obviously we cannot do both in a single value
IMO keeping integers is more important here as there is real data loss; + vs - 0 does not change the value - moreover the issue only happens for `+0` and `-0` that have no decimal point, it is all preserved well for `+0.0` and `-0.0`.

radeusgd committed:
fix problem aggregation losing some warnings

radeusgd committed:
special case to keep large numbers as integers, but still be able to parse `-0` as `-0.0` in decimal mode - remembering the zero-sign
